File: app/src/main/java/com/vikingsen/cheesedemo/model/data/cheese/CheeseRepository.java
package com.vikingsen.cheesedemo.model.data.cheese;


import com.vikingsen.cheesedemo.model.database.cheese.Cheese;

import org.dbtools.android.domain.DatabaseTableChange;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Maybe;
import io.reactivex.Observable;
import io.reactivex.Single;
import kotlin.Pair;
import timber.log.Timber;

@Singleton
public class CheeseRepository {

    private final CheeseRemoteDataSource remoteDataSource;
    private final CheeseLocalDataSource localDataSource;

    @Inject
    CheeseRepository(CheeseRemoteDataSource remoteDataSource, CheeseLocalDataSource localDataSource) {
        this.remoteDataSource = remoteDataSource;
        this.localDataSource = localDataSource;
    }

    public Single<List<Cheese>> getCheeses(boolean forceRefresh) {
        return Single.just(forceRefresh)
                .flatMap(force -> {
                    Single<List<Cheese>> remoteCheese = getAndSaveRemoteCheeses();
                    Single<List<Cheese>> localCheese = localDataSource.getCheeses();
                    if (force || localDataSource.isCheeseStale()) {
                        Timber.d("CHEESE IS STALE");
                        return concatSources(remoteCheese, localCheese);
                    } else {
                        Timber.d("CHEESE IS FRESH");
                        return concatSources(localCheese, remoteCheese);
                    }
                });
    }

    public Maybe<Cheese> getCheese(long cheeseId, boolean forceRefresh) {
        return Maybe.just(new Pair<>(cheeseId, forceRefresh))
                .flatMap(pair -> {
                    long id = pair.getFirst();
                    boolean force = pair.getSecond();
                    Maybe<Cheese> remoteCheese = getAndSaveRemoteCheese(id);
                    Maybe<Cheese> localCheese = localDataSource.getCheese(id);
                            if (force || localDataSource.isCheeseStale(id)) {
                                Timber.d("CHEESE IS STALE");
                                return concatSources(remoteCheese, localCheese);
                            } else {
                                Timber.d("CHEESE IS FRESH");
                                return concatSources(localCheese, remoteCheese);
                            }
                        }
                );
    }

    public long modelTs() {
        return localDataSource.modelTs();
    }

    public Observable<DatabaseTableChange> dataChanges() {
        return localDataSource.dataChanges();
    }

    public boolean isUpdatedSince(long modelTs) {
        return modelTs() != modelTs;
    }

    private Single<List<Cheese>> getAndSaveRemoteCheeses() {
        return remoteDataSource.getCheeses().map(cheeses -> localDataSource.saveCheeses(cheeses));
    }

    private Maybe<Cheese> getAndSaveRemoteCheese(Long cheeseId) {
        return remoteDataSource.getCheese(cheeseId).map(cheese -> localDataSource.saveCheese(cheese));
    }

    private Single<List<Cheese>> concatSources(Single<List<Cheese>> source1, Single<List<Cheese>> source2) {
        return Single.concat(source1, source2)
                .filter(list -> !list.isEmpty())
                .first(Collections.emptyList());
    }

    private Maybe<Cheese> concatSources(Maybe<Cheese> source1, Maybe<Cheese> source2) {
        return Maybe.concat(source1, source2)
                .firstElement();
    }
}

File: app/src/main/java/com/vikingsen/cheesedemo/model/data/cheese/CheeseRepository.kt
package com.vikingsen.cheesedemo.model.data.cheese

import android.arch.lifecycle.LiveData
import com.vikingsen.cheesedemo.model.NetworkBoundResource
import com.vikingsen.cheesedemo.model.Resource
import com.vikingsen.cheesedemo.model.database.cheese.Cheese
import com.vikingsen.cheesedemo.model.webservice.dto.CheeseDto
import com.vikingsen.cheesedemo.util.CoroutineContextProvider
import org.threeten.bp.LocalDateTime
import org.threeten.bp.temporal.ChronoUnit
import javax.inject.Inject
import javax.inject.Singleton


@Singleton
class CheeseRepository
@Inject constructor(
        private val remoteDataSource: CheeseRemoteDataSource,
        private val localDataSource: CheeseLocalDataSource,
        private val ccp: CoroutineContextProvider
) {

    fun getCheeses(forceRefresh: Boolean): LiveData<Resource<List<Cheese>>> =
            object : NetworkBoundResource<List<Cheese>, List<CheeseDto>>(ccp.commonPool) {
                override fun loadFromDb() = localDataSource.getCheeses()

                suspend override fun shouldFetch(data: List<Cheese>?): Boolean {
                    if (forceRefresh || data == null) {
                        return true
                    }

                    val cacheExpiration = LocalDateTime.now().minus(CACHE_VALID_AMOUNT, CACHE_VALID_UNIT)
                    // GOTCHA - DOUBLE CHECK THAT THE CHECK MATCHES THE METHOD NAME (isBefore vs isAfter)
                    return data.map { it.cached }.min()?.isBefore(cacheExpiration) ?: true
                }

                suspend override fun fetchFromNetwork(): NetworkResponse<List<CheeseDto>> {
                    return remoteDataSource.getCheeses()
                }

                suspend override fun saveNetworkData(data: List<CheeseDto>) {
                    localDataSource.saveCheeses(data)
                }

            }.asLiveData()

    fun getCheese(cheeseId: Long, forceRefresh: Boolean): LiveData<Resource<Cheese>> =
            object : NetworkBoundResource<Cheese, CheeseDto>(ccp.commonPool) {
                override fun loadFromDb() = localDataSource.getCheese(cheeseId)

                suspend override fun shouldFetch(data: Cheese?): Boolean {
                    if (forceRefresh || data == null) {
                        return true
                    }

                    val cacheExpiration = LocalDateTime.now().minus(CACHE_VALID_AMOUNT, CACHE_VALID_UNIT)
                    // GOTCHA - DOUBLE CHECK THAT THE CHECK MATCHES THE METHOD NAME (isBefore vs isAfter)
                    return data.cached.isBefore(cacheExpiration)
                }

                suspend override fun fetchFromNetwork(): NetworkResponse<CheeseDto> {
                    return remoteDataSource.getCheese(cheeseId)
                }

                suspend override fun saveNetworkData(data: CheeseDto) {
                    localDataSource.saveCheese(data)
                }
            }.asLiveData()

    companion object {
        private val CACHE_VALID_AMOUNT = 1L
        private val CACHE_VALID_UNIT = ChronoUnit.HOURS
    }
}